Added comment to Decoder.finalize

smilu97 opened the pull request:
Hi. I'm Korean student, recently studying tensorflow.

At first, I'm sorry about my poor english.

I found a curious thing that only `finalize` method in `contrib.seq2seq.Decoder` doesn't have any comment in documents.

So, I tried to write some comments on it, usually copying from comments from other methods.

smilu97 committed:
Added comment to Decoder.finalize

smilu97 commented:
I signed the indivisual CLA!

ebrevdo commented on tensorflow/contrib/seq2seq/python/ops/decoder.py:
```diff
@@ -112,6 +112,20 @@ def step(self, time, inputs, state, name=None):
+        steps.
+      final_state: RNNCell final state (possibly nested tuple of) tensor[s] for
+        last time step.
+      sequence_lengths: Scalar `int32` tensor. Maximum sequence length.
```
I don't think this is the maximum.  It is a vector of lengths

smilu97 replied:
```python3
#!/usr/bin/env python
'''
Seq2Seq implementation, practice
Example of tf.contrib.dynamic_decode
'''

import numpy as np
import tensorflow as tf
import tensorflow.contrib.seq2seq as seq2seq

class Decoder(seq2seq.Decoder):

    def __init__(self, batch_size, initial_input, initial_state):
        self._batch_size = batch_size
        self._initial_input = initial_input
        self._initial_state = initial_state

    @property
    def batch_size(self):
        return self._batch_size
    
    @property
    def output_dtype(self):
        return tf.float32
    
    @property
    def output_size(self):
        return 3
    
    @property
    def tracks_own_finished(self):
        return True

    def finalize(self, outputs, final_state, sequence_lengths):
        final_outputs = outputs
        final_state = final_state
        print('finalize: sequence_lengths:', sequence_lengths)
        return final_outputs, final_state
    
    def initialize(self, name=None):
        finished = False
        initial_inputs = self._initial_input
        initial_state = self._initial_state
        return finished, initial_inputs, initial_state
    
    def step(self, time, inputs, state, name=None):
        outputs = state + inputs
        next_state = outputs
        next_inputs = outputs
        finished = tf.reduce_sum(outputs) >= 100
        return outputs, next_state, next_inputs, finished

def main():

    initial_input = tf.constant([[1,2,3], [2,3,4]], dtype=tf.float32)
    initial_state = tf.constant([[1,2,3], [2,3,4]], dtype=tf.float32)
    decoder = Decoder(2, initial_input, initial_state)
    outputs, state, seq_lengths = seq2seq.dynamic_decode(decoder)

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        print('outputs:', sess.run(outputs))

if __name__ == '__main__':
    main()
```

I tried to make a sample code to know what is the `sequence_lengths`, the argument of `finalize`.

Below is result of code-run
```
finalize: sequence_lengths: Tensor("decoder/while/Exit_5:0", shape=(), dtype=int32)
outputs: [[[ 2.  4.  6.]
  [ 4.  8. 12.]
  [ 8. 16. 24.]]

 [[ 4.  6.  8.]
  [ 8. 12. 16.]
  [16. 24. 32.]]]
```
Above shows `sequence_lengths.shape` is ()

I think there may be other options to make finalize get `sequence_lengths` as ndarray of `int32`, but I can't find it.

If you don't mind, can you give me some hint, or help?

smilu97 replied:
I apologize you if you wasted your time to read above reply...

I found the way to get `sequence_lengths` as ndarray of `int32`.

I'm sorry

smilu97 committed:
Fix describe, argument `sequence_lengths` in seq2seq.Decoder.finalize

smilu97 committed:
clarify dim_size of sequence_lengths of `seq2seq.Decoder.finalize`

smilu97 commented:
I changed comment of `sequence_lengths`

tensorflowbutler commented:
Nagging Reviewer @ebrevdo, @lmthang: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @ebrevdo, @lmthang: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @ebrevdo, @lmthang: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @ebrevdo, @lmthang: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 59 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @ebrevdo, @lmthang: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 74 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @ebrevdo, @lmthang: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

ebrevdo commented on tensorflow/contrib/seq2seq/python/ops/decoder.py:
```diff
@@ -112,6 +112,20 @@ def step(self, time, inputs, state, name=None):
     raise NotImplementedError
 
   def finalize(self, outputs, final_state, sequence_lengths):
+    """Called after any decoding iterations.
```
Called after decoding iterations complete.

ebrevdo commented on tensorflow/contrib/seq2seq/python/ops/decoder.py:
```diff
@@ -112,6 +112,20 @@ def step(self, time, inputs, state, name=None):
+        steps.
+      final_state: RNNCell final state (possibly nested tuple of) tensor[s] for
+        last time step.
+      sequence_lengths: 1-D `int32` tensor containing lengths of each sequences.
```
... of each sequence.

smilu97 replied:
I changed. sorry for bad grammar

smilu97 committed:
Update decoder.py as discussion

https://github.com/tensorflow/tensorflow/pull/24195#discussion_r277175548
https://github.com/tensorflow/tensorflow/pull/24195#discussion_r277175539

gbaned commented:
@smilu97 Hi, could you please look into the changes requested by ebrevdo ?

smilu97 commented:
Oh.. I already commited new one as ebrevdo requested, but I forgot to write some comment about this.
or did I missed something?

smilu97 committed:
Resolve C0301[line too long] in ops/decoder:127

smilu97 commented:
This failed pylint at C0301[line-too-long], so I fixed some
